src/backend/cuda/kernel/random_philox.hpp

```diff
@@ -0,0 +1,394 @@
+    static const uintl w64_1 = uintl(0xBB67AE8584CAA73B);
+    static const uint w32_0 = uint(0x9E3779B9);
+    static const uint w32_1 = uint(0xBB67AE85);
+    static const unsigned int PhiloxDefaultRounds = 10;
+    #define UINTMAXFLOAT 4294967296.0f
+    #define UINTLMAXDOUBLE 4294967296.0*4294967296.0
+
+    static inline __device__ void mulhilo(const uint &a, const uint &b,
+            uint &hi, uint &lo)
+    {
+        hi = __umulhi(a,b);
+        lo = a*b;
+    }
+
+    static inline __device__ void mulhilo(const uintl &a, const uintl &b,
+            uintl &hi, uintl &lo)
+    {
+        hi = __umul64hi(a,b);
+        lo = a*b;
+    }
+
+    template <typename T, T mul>
+    static inline __device__ void phRound(PhiloxCounter<T, 2> &ctr,
+            const PhiloxKey<T, 2> &key)
+    {
+        T hi, lo;
+        mulhilo(mul, ctr.v[0], hi, lo);
+        ctr.v[0] = hi^key.v[0]^ctr.v[1];
+        ctr.v[1] = lo;
+    }
+
+    template <typename T, T mul0, T mul1>
+    static inline __device__ void phRound(PhiloxCounter<T, 4> &ctr,
+            const PhiloxKey<T, 4> &key)
+    {
+        T hi0, lo0, hi1, lo1;
+        mulhilo(mul0, ctr.v[0], hi0, lo0);
+        mulhilo(mul1, ctr.v[2], hi1, lo1);
+        ctr.v[0] = hi1^ctr.v[1]^key.v[0];
+        ctr.v[1] = lo1;
+        ctr.v[2] = hi0^ctr.v[3]^key.v[1];
+        ctr.v[3] = lo0;
+    }
+
+    static inline __device__ void philoxRound(PhiloxCounter<uint, 4> &ctr,
+            const PhiloxKey<uint, 4> &key)
+    {
+        phRound<uint, m4x32_0, m4x32_1>(ctr, key);
+    }
+
+    static inline __device__ void philoxRound(PhiloxCounter<uintl, 4> &ctr,
+            const PhiloxKey<uintl, 4> &key)
+    {
+        phRound<uintl, m4x64_0, m4x64_1>(ctr, key);
+    }
+
+    static inline __device__ void philoxRound(PhiloxCounter<uint, 2> &ctr,
+            const PhiloxKey<uint, 2> &key)
+    {
+        phRound<uint, m2x32_0>(ctr, key);
+    }
+
+    static inline __device__ void philoxRound(PhiloxCounter<uintl, 2> &ctr,
+            const PhiloxKey<uintl, 2> &key)
+    {
+        phRound<uintl, m2x64_0>(ctr, key);
+    }
+
+    static inline __device__ void philoxBump(PhiloxKey<uint, 2> &key)
+    {
+        key.v[0] += w32_0;
+    }
+
+    static inline __device__ void philoxBump(PhiloxKey<uintl, 2> &key)
+    {
+        key.v[0] += w64_0;
+    }
+
+    static inline __device__ void philoxBump(PhiloxKey<uint, 4> &key)
+    {
+        key.v[0] += w32_0;
+        key.v[1] += w32_1;
+    }
+
+    static inline __device__ void philoxBump(PhiloxKey<uintl, 4> &key)
+    {
+        key.v[0] += w64_0;
+        key.v[1] += w64_1;
+    }
+
+    template <typename T, unsigned int W, unsigned int R>
+    static inline __device__ PhiloxCounter<T,W> philox(const PhiloxCounter<T,W> &inputCtr,
+            const PhiloxKey<T,W> &inputKey)
+    {
+        PhiloxCounter<T,W> ctr = inputCtr;
+        PhiloxKey<T,W> key = inputKey;
+        if (R>0)    {                   philoxRound(ctr, key);}
```

Comment:
@kaatish If `R` is going to be compile time constant, you can use a for loop here with `#pragma unroll` on top of it. It will have a performance close to this one.